nixpkgs-manual: Document `runCommandWith`, refactor `runCommand{,CC,Local}` (#333262)
NixOS/nixpkgs: doc/build-helpers/trivial-build-helpers.chapter.md

Nixpkgs provides a variety of wrapper functions that help build commonly useful derivations.
Like [`stdenv.mkDerivation`](#sec-using-stdenv), each of these build helpers creates a derivation, but the arguments passed are different (usually simpler) from those required by `stdenv.mkDerivation`.

## `runCommandWith` {#trivial-builder-runCommandWith}

`runCommandWith :: { name, runLocal ? false, stdenv ? pkgs.stdenv, derivationArgs } -> String -> Derivation`

`runCommandWith args buildCommand` produces a derivation which is built using the specified `buildCommand`.

It is a generalized version of `runCommand*`, whose behaviour is controlled via a single attribute set passed
as the first argument, and allows specifying `stdenv` freely.

`name :: String`
:   The derivation's name, which Nix will append to the store path; see [`mkDerivation`](#sec-using-stdenv).

`runLocal :: Bool`

> :   Whether to forces the derivation to be built locally, forgoing the use of substitutors and distributed builds.

Reviewer: I find this sentence a little hard to grasp.
```suggestion
:   If set to `true` this forces the derivation to be built locally. Remote substitutes and distributed builds, won't be used.
```

    This is intended for very cheap commands (<1s execution time) which can be sped up by avoiding the network round-trip(s).
    Its effect is to set [`preferLocalBuild = true`][preferLocalBuild] and [`allowSubstitutes = false`][allowSubstitutes].

   ::: {.note}

>    This prevents the use of substitutors, so only set `runLocal` (or use `runCommandLocal`) when certain the user will

Reviewer: Can we link to what "substitutors" are

Nixpkgs provides a variety of wrapper functions that help build commonly useful derivations.
Like [`stdenv.mkDerivation`](#sec-using-stdenv), each of these build helpers creates a derivation, but the arguments passed are different (usually simpler) from those required by `stdenv.mkDerivation`.

## `runCommandWith` {#trivial-builder-runCommandWith}

> The function `runCommand` returns a derivation built using the specified command(s), in a specified environment.

Reviewer: ```suggestion
The function `runCommandWith` returns a derivation built using the specified command(s), in a specified environment.
```

> It is a generalized version of `runCommand*`, whose behaviour is controlled via a single attribute set passed

Reviewer: ```suggestion
It is the underlying base function of  all `runCommand*` variants. 
The general behavior is controlled via a single attribute set passed 
```

> as the first argument, and allows specifying `stdenv` freely.

Reviewer: ```suggestion
as the first argument, and allows specifying `stdenv` freely.

The following `runCommand*` variants exist: 
- [runCommand](#trivial-builder-runCommand})
- (provide cross references here)
- ...
```

Reviewer: We should do cross references, so it is easier to go back and forth between the functions.
With that i mean links that go in both directions.
runCommandWith <-> runCommand

> ### Type

Reviewer: > >      failed to render manual manual.md
> >
> > caused by:
> >         processing included file build-helpers.md from line 12
> >
> > caused by:
> >    processing included file build-helpers/trivial-build-helpers.chapter.md from line 21
> >
> > caused by:
> >      heading in line 14 does not have an id
> 
> Did I get the syntax wrong or is the correct thing to do, manually stuffing a unique anchor id on each type and inputs section?

Yes. Unfortunately the id must be handwritten currently. This was introduced to guarantee stable reference links

`runLocal` (Boolean)
:   If set to `true` this forces the derivation to be built locally, not using [substitutes] nor remote builds.
    This is intended for very cheap commands (<1s execution time) which can be sped up by avoiding the network round-trip(s).
    Its effect is to set [`preferLocalBuild = true`][preferLocalBuild] and [`allowSubstitutes = false`][allowSubstitutes].

   ::: {.note}
   This prevents the use of [substituters][substituter], so only set `runLocal` (or use `runCommandLocal`) when certain the user will
   always have a builder for the `system` of the derivation. This should be true for most trivial use cases
   (e.g., just copying some files to a different location or adding symlinks) because there the `system`
   is usually the same as `builtins.currentSystem`.
   :::

`stdenv` (Derivation)
:   The [standard environment](#chap-stdenv) to use, defaulting to `pkgs.stdenv`

`derivationArgs` (Attribute set)
:   Additional arguments for [`mkDerivation`](#sec-using-stdenv).

<!-- TODO(nicoo) how to define that this is `runCommandWith`'s second argument? -->

> `buildCommand` (String)

Reviewer 2: I think your choice to do it like this is just fine. The word positional comes to mind.

:   Shell commands to run in the derivation builder.

    ::: {.note}
    You have to create a file or directory `$out` for Nix to be able to run the builder successfully.
    :::

[allowSubstitutes]: https://nixos.org/nix/manual/#adv-attr-allowSubstitutes
[preferLocalBuild]: https://nixos.org/nix/manual/#adv-attr-preferLocalBuild
[substituter]: https://nix.dev/manual/nix/latest/glossary#gloss-substituter
[substitutes]: https://nix.dev/manual/nix/2.23/glossary#gloss-substitute

::: {.example #ex-runcommandwith}
# Invocation of `runCommandWith`

```nix
runCommandWith {
  name = "example";
  derivationArgs.nativeBuildInputs = [ cowsay ];
} ''
  cowsay > $out <<EOMOO
  'runCommandWith' is a bit cumbersome,

>   so we have more-ergonomic wrappers.

Reviewer 2: ```suggestion
  so we have more ergonomic wrappers.
```

The function `runCommand` returns a derivation built using the specified command(s), in the `stdenvNoCC` environment.

`runCommandCC` is similar but uses the default compiler environment. To minimize dependencies, `runCommandCC`
should only be used when the build command needs a C compiler.

`runCommandLocal` is also similar to `runCommand`, but forces the derivation to be built locally.
See the note on [`runCommandWith`] about `runLocal`.

[`runCommandWith`]: #trivial-builder-runCommandWith

### Type {#trivial-builder-runCommand-Type}

```
runCommand      :: String -> AttrSet -> String -> Derivation
runCommandCC    :: String -> AttrSet -> String -> Derivation
runCommandLocal :: String -> AttrSet -> String -> Derivation
```

### Input {#trivial-builder-runCommand-Input}

While the type signature(s) differ from [`runCommandWith`], individual arguments with the same name will have the same type and meaning:

`name` (String)
:   The derivation's name

`derivationArgs` (Attribute set)
:   Additional parameters passed to [`mkDerivation`]

`buildCommand` (String)
:   The command(s) run to build the derivation.

::: {.example #ex-runcommand-simple}
# Invocation of `runCommand`

```nix

> (import <nixpkgs> {}).runCommand "my-example" {} ''

Reviewer 2: Glad to see this gone.

    You have to create a file or directory `$out` for Nix to be able to run the builder successfully.
    :::

[allowSubstitutes]: https://nixos.org/nix/manual/#adv-attr-allowSubstitutes
[preferLocalBuild]: https://nixos.org/nix/manual/#adv-attr-preferLocalBuild

::: {.example #ex-runcommandwith}
# Invocation of `runCommandWith`

```nix
runCommandWith {
  name = "example";
  derivationArgs.nativeBuildInputs = [ cowsay ];
} ''
  cowsay > $out <<EOMOO
  'runCommandWith' is a bit cumbersome,
  so we have more-ergonomic wrappers.
  EOMOO
''
```

:::

## `runCommand` and `runCommandCC` {#trivial-builder-runCommand}

`runCommand :: String -> AttrSet -> String -> Derivation`
`runCommandCC :: String -> AttrSet -> String -> Derivation`

> `runCommand name derivationArgs buildCommand` is a derivation which runs the specified `buildCommand`

Reviewer: `runCommand` and `runCommandCC` are functions. We have some conventions in `lib` how to document functions.
-> https://github.com/NixOS/nixpkgs/blob/master/lib/asserts.nix
This is a little inconsistent.

Reviewer: Don't look at other `pkgs.*` function documentation in the manual. We are currently refactoring a lot of things.

You should follow this guide: (starting at: "When documenting functions ...")
https://github.com/NixOS/nixpkgs/blob/667f3a73f7ed070b570c03d63275acc301e7de54/doc/README.md#documentation-conventions

All functions that you encountered and don't follow those conventions yet should be/ will be refactored.

Reviewer: ```suggestion
The function `runCommand` returns a derivation which runs the specified `buildCommand` when beeing built.
```

using the `stdenvNoCC` environment.  `runCommandCC` is similar but uses the default compiler environment.

To minimize dependencies, `runCommandCC` should only be used when `buildCommand` needs a C compiler.

> The arguments `name`, `derivationArgs`, and `buildCommand` have the same types and meaning as for [`runCommandWith`](#trivial-builder-runCommandWith)

Reviewer: It would be better to assume that the reader doesn't need to jump to `runCommandWith` first.

Reviewer: ```suggestion
The arguments `name`, `derivationArgs`, and `buildCommand` have the same types and meaning as for [`runCommandWith`](#trivial-builder-runCommandWith)

arguments are inherited from [`runCommandWith`](#trivial-builder-runCommandWith)
```

Also we should document the inputs of the function (arguments) following the definition list convention that i posted above.

```
### Inputs

name (String)
: name of the produced derivation
env (AttrSet)
: Additional attributes passed to underlying mkDerivation
buildCommand (String)
: Command that is executed at build time.
```

```
:::

::: {.note}
`runCommand name derivationArgs buildCommand` is equivalent to
```nix
runCommandWith {
  inherit name derivationArgs;
  stdenv = stdenvNoCC;
} buildCommand
```

Likewise, `runCommandCC name derivationArgs buildCommand` is equivalent to
```nix
runCommandWith {
  inherit name derivationArgs;
} buildCommand
```

:::

## `runCommandLocal` {#trivial-builder-runCommandLocal}

> `runCommandLocal :: String -> AttrSet -> String -> Derivation`

Reviewer: ````suggestion
# Type
```
runCommandLocal :: String -> { <type out please> }-> String -> Derivation``
```
````

This is more precise and matches what we have in lib already for all other functions.

The type `AttrSet` doesnt add much value. I'd prefer something like `{ derivationArgs :: { ... }; stdenv :: Derivation; ... }` because that is more concrete

Reviewer: When i look at the function definition of `runCommandWith` i see the following:

```nix
  runCommandWith =
    let
      # prevent infinite recursion for the default stdenv value
      defaultStdenv = stdenv;
    in
    {
      # which stdenv to use, defaults to a stdenv with a C compiler, pkgs.stdenv
      stdenv ? defaultStdenv
      # whether to build this derivation locally instead of substituting
    , runLocal ? false
      # extra arguments to pass to stdenv.mkDerivation
    , derivationArgs ? { }
      # name of the resulting derivation
    , name
      # TODO(@Artturin): enable strictDeps always
    }: buildCommand:
```

This would have the following type:

```
runCommandLocal :: String -> {
    name :: name;
    stdenv? :: Derivation;
    runLocal? :: Bool;
    derivationArgs? :: { ... };  <- args forwarded to mkDerivation
  }-> String -> Derivation``
```

If you could use this instead of Attrs i think would be much better.

Reviewer: Oh sorry for the confusion. This comment was intended for `runCommandWith`
